### ISSUE TITLE
`stringify` should throw when serializing a BigInt like `JSON.stringify` does

### ISSUE THREAD

**Issue by jordanbtucker** [json5/json5#273]
`stringify` should throw an error when serializing a BigInt to match JSON behavior.

```js
const json = JSON.stringify(123n)
// Uncaught TypeError: Do not know how to serialize a BigInt

const json5 = JSON5.stringify(123n)
// json5 == undefined
```

In order to serialize a BigInt, one could use a replacer function that converts a BigInt to either a number or a string. Alternatively, we could add an option to `stringify` that enables BigInt support.

**Comment by chipweinberger** [comment]
JSON5 nor JSON document a maximum number size. IMO, JSON5 should support arbitrarily large numbers.

**Comment by jordanbtucker** [comment]
@chipweinberger I [agree](https://github.com/json5/json5-spec/issues/36#issuecomment-1139787561) that the JSON5 format should continue to allow for arbitrary-precision integers as it currently does.

However, JSON also allows for arbitrary-precision integers, ECMAScript 2020 supports BigInt and BigInt literals, yet Node.js and browsers still throw an error when stringifying BigInt values with JSON. My question is, why? Usually there are specific reasons for this kind of behavior. I'd like to see if we can find that out before deciding which way to go.

**Comment by jcubic** [comment]
@jordanbtucker the reason why `JSON.stringify` throws an error is that BigInt is not part of JSON spec. You can't serialize BigInt because it doesn't have a literal representation in JSON. But JSON5 should support BigInt, it can just create bigInt literal. I don't see a reason why JSON5 should throw an error. JSON5 is JSON that is more like JavaScript, and JS supports BigInt literals.

**Comment by jordanbtucker** [comment]
@jcubic Both JSON and JSON5 are capable of representing BigInts because neither spec sets limits on the precision of integers. Both [RFC 8259](https://www.rfc-editor.org/rfc/rfc8259#section-6) and [RFC 7493](https://www.rfc-editor.org/rfc/rfc7493#section-2.2) recommend sticking to numbers that can be represented in IEEE 754 for interoperability.

The latest [ES spec](https://tc39.es/ecma262/#sec-serializejsonproperty) explicitly throws an error when stringifying BigInt, not because JSON doesn't support BigInt, but because the existing JSON implementations rely on numbers being IEEE 754 floating-points. JSON5 is in a similar situation. If we start silently supporting BigInt, then that could create interoperability problems for other JSON5 parsers. Also keep in mind that neither JSON nor JSON5 are only used with JavaScript, and not all environments support arbitrary-precision integers.

If a user wants to stringify BigInt, then they can add a `toJSON` method to `BigInt.prototype` or use a `replacer` function. This currently works with JSON and JSON5. Throwing on stringifying BigInts will keep JSON5 behavior in line with JSON behavior for interoperability reasons.

I recommend reading the discussions in https://github.com/tc39/proposal-bigint/issues/24 and https://github.com/json5/json5-spec/issues/36

**Comment by jordanbtucker** [comment]
I should mention that adding a `toJSON` method to `BigInt.prototype` or using a `replacer` function will not allow a user to create a JSON or JSON5 document that contains numbers outside the range of IEEE 754 floating-points. You still need to use a string for numbers outside that range.

**Comment by jcubic** [comment]
Thanks, I will try to use a replacer in my project for BitInt. I've just realized that I was using serialization of custom classes with a replacer in my other project but I've replaced it with [CBOR](https://github.com/kriszyp/cbor-x).

**Comment by jordanbtucker** [comment]
I'm not opposed to adding an option that would allow `stringify` to generate JSON5 documents with large integers using BigInt, but this would be an opt-in feature, there would be no `n` suffix, and the documentation should mention the interoperability implications.

For `parse`, an option could be added to parse integers greater than `Number.MAX_SAFE_INTEGER` or less than `Number.MIN_SAFE_INTEGER` as BigInts.

**Comment by jordanbtucker** [comment]
Fixed in 4af0e1c0ff4c662cc68024f1581aaff864e96067
